
---

# 🌟 𝐇𝐲𝐩𝐨𝐭𝐡𝐞𝐬𝐢𝐬 𝐓𝐞𝐬𝐭𝐢𝐧𝐠 🌟

🔍 **Hypothesis testing** is a **statistical method** used to **make decisions** about **population data** based on **sample data**.  
  
📊 It **evaluates** whether there is **enough evidence** in the sample to **support** a **specific claim** (hypothesis) about the **population**.

✅ If the evidence is strong enough, we **accept** the claim;  
❌ if not, we **reject** it and look for other explanations!

---

# 🧠 𝐖𝐡𝐞𝐧 𝐭𝐨 𝐠𝐨 𝐟𝐨𝐫 𝐇𝐲𝐩𝐨𝐭𝐡𝐞𝐬𝐢𝐬 𝐓𝐞𝐬𝐭𝐢𝐧𝐠?

👉 When you want to **verify** or **prove a statement** about a population.  
👉 When you have a **sample** and need to **infer** about the **whole population**.  
👉 When you need to **test a claim** scientifically (example: "Our new product increases battery life! 🔋").

**Typical examples:**  
- Testing if a new drug is effective 💊  
- Checking if a machine produces items within tolerance 🎯  
- Comparing two teaching methods 🏫

---

# 🚫 𝐈𝐟 𝐲𝐨𝐮 𝐝𝐨𝐧'𝐭 𝐰𝐚𝐧𝐭 𝐭𝐨 𝐮𝐬𝐞 𝐇𝐲𝐩𝐨𝐭𝐡𝐞𝐬𝐢𝐬 𝐓𝐞𝐬𝐭𝐢𝐧𝐠...

🔵 You can use **Estimation Methods** instead:
  
📈 **Confidence Intervals**:  
- Instead of "proving" something, you "estimate" the range where the true population value likely falls.
- Example: "We are 95% confident the true average delivery time is between 29 and 31 minutes."

📉 **Descriptive Statistics**:  
- Just **describe** the data using mean, median, standard deviation, etc.
- No formal "testing", just **summarizing** the data nicely.

---

# 🎯 𝐒𝐮𝐦𝐦𝐚𝐫𝐲 🎯

| Situation                              | Use Hypothesis Testing?  | Alternative |
|:---------------------------------------|:------------------------:|:-----------:|
| Need to **prove a claim** 📢            | ✅ Yes                   | -           |
| Just want to **estimate** 🎯            | 🚫 No                    | Confidence Intervals |
| Just want to **describe data** 📊       | 🚫 No                    | Descriptive Statistics |

---



# 🌟 𝐓𝐲𝐩𝐞𝐬 𝐨𝐟 𝐇𝐲𝐩𝐨𝐭𝐡𝐞𝐬𝐢𝐬 𝐓𝐞𝐬𝐭𝐢𝐧𝐠 🌟


---

| ✅ Situation | 🧪 Use **T-test** | 🧪 Use **Z-test** |
|:--|:--|:--|
| Standard Deviation σ | ❌ Unknown | ✅ Known |
| Sample Size | 📉 Small (n < 30) | 📈 Large (n ≥ 30) |
| Distribution | ✅ Normal or nearly normal | ✅ Normal or large enough for CLT |

---



---

# 🌟 **T-Tests and Hypothesis Statements** 🌟

---

## 1. **One-Sample t-Test** 🧑‍🏫

**Used when:**  
You have a **single sample** and want to compare its **mean** to a known **population mean**.

**Example:**  
You have a sample of 25 students' test scores, and you want to know if the **average score** differs from the population mean of 75.

### Hypothesis:
- **Null Hypothesis (H₀):** The sample mean is equal to 75.  
  \[
  H_0: mu = 75
  \]
- **Alternative Hypothesis (H₁):** The sample mean is not equal to 75.  
  \[
  H_1: mu != 75
  \]

---



In [1]:
# ONE SAMPLE T-TEST
marks = [11,17,18,14,19,19.5,18,16,13,17,16,14,15,18,19,11,10,7,2,9,3,14,17,16,19]
population_mean = 75

# Hypotheses
ho = "sample mean = population mean"
ha = "sample mean != population mean"

# Perform one-sample t-test
from scipy import stats
t_statistic, p_value = stats.ttest_1samp(marks, population_mean)

print("t-statistic:", t_statistic)
print("p-value:", p_value)

# Significance level
ALPHA = 0.05

# Decision
if p_value < ALPHA:
    print(f"Reject null hypothesis. {ha}")
else:
    print(f"Fail to reject null hypothesis. {ho}")


t-statistic: -62.64716008460932
p-value: 4.108306354841194e-28
Reject null hypothesis. sample mean != population mean


## 2. **Two-Sample t-Test** 🆚

**Used when:**  
You want to compare the **means** of **two independent samples** to see if they are **significantly different**.

**Example:**  
You want to compare the average height of male and female students in a class.

### Hypothesis:
- **Null Hypothesis (H₀):** The mean height of males is equal to the mean height of females.  
  [
  H_0: mu_1 = mu_2
  ]
- **Alternative Hypothesis (H₁):** The mean height of males is not equal to the mean height of females.  
  [
  H_1: mu_1 != mu_2
  ]

---



In [2]:
# TWO SAMPLE T-TEST:
# TWO SAMPLES ARE GIVE SAID TO COMPARE BOTH SAMPLES MEAN

MALE_FOOTS=[4,5,6,7,4,5,6,7,5,6,7,5.8,6.3,5.7,5.5,5.3,4.9]
FEMALE_FOOTS=[3.8,4.2,4.8,4.9,4.5,5.2,5.1,4,5.4,4.6,4,4.9,5.5,5,5.3,5.9,6]
MU1=sum(MALE_FOOTS)/len(MALE_FOOTS)
MU2=sum(FEMALE_FOOTS)/len(FEMALE_FOOTS)

HO="MALE HEIGHT MEAN IS EQUAL TO FEMALE HIEGHT MEAN"
HA="MALE HEIGHT MEAN IS NOT EQUAL TO FEMALE HEIGHT MEAN"

from scipy import stats

t_statistic, p_value = stats.ttest_ind(MALE_FOOTS, FEMALE_FOOTS, equal_var=False)  # assume equal variances

print("t-statistic:", t_statistic)
print("p-value:", p_value)


# COMPARISON OF P_VALUE AND ALPHA:
ALPHA=0.05

if p_value<ALPHA:
    print(f"reject null hypothesis {HA}")
else:
    print(f"accept null hypothesis {HO}")    


t-statistic: 2.6645879343772574
p-value: 0.012525132809984007
reject null hypothesis MALE HEIGHT MEAN IS NOT EQUAL TO FEMALE HEIGHT MEAN


## 3. **One-Sample One-Tailed t-Test** (Right or Left) ➡️ ⬅️

**Used when:**  
You test if the sample mean is either **greater than** or **less than** the population mean. It tests only in **one direction**.

### Right-Tailed Test (Greater Than):
**Example:**  
You want to test if the average temperature in a city is **greater** than 30°C.

### Hypothesis:
- **Null Hypothesis (H₀):** The mean temperature is less than or equal to 30°C.  
  [
  H_0: mu < 30
  ]
- **Alternative Hypothesis (H₁):** The mean temperature is **greater** than 30°C.  
  [
  H_1: mu > 30
  ]

### Left-Tailed Test (Less Than):
**Example:**  
You want to test if the average temperature in a city is **less** than 30°C.

### Hypothesis:
- **Null Hypothesis (H₀):** The mean temperature is greater than or equal to 30°C.  
  [
  H_0: mu > 30
  ]
- **Alternative Hypothesis (H₁):** The mean temperature is **less** than 30°C.  
  [
  H_1: mu < 30
  ]

---



In [3]:
# ONE SAMPLE ONE TAILED TEST:
# Given data:
temperatures = [29.5, 30.1, 30.3, 30.8, 31.2, 30.5, 29.9, 31.5, 30.7, 30.9,
                31.0, 30.6, 30.2, 30.3, 31.4, 31.6, 30.8, 30.9, 31.1, 30.4]

# Population mean to compare with:
population_mean = 30

# stats the hypothesis:
ho="sample mean > 30"
ha="sample mean< 30"

# Perform one-sample TWO TAILED t-test... 
from scipy import stats
t_stat, p_value_two_tailed = stats.ttest_1samp(temperatures, population_mean)

# CONVERT P_VALUE_TWO_TAILED TO ONE TAIL P_VALUE:
p_value=p_value_two_tailed/2


# Print results:
import numpy as np
print("population_mean :",population_mean)
print("Sample Mean:", np.mean(temperatures))
print("t-statistic:", t_stat)
print("p-value (one-tailed):", p_value)

# Significance level:
alpha = 0.05   

# Decision:
if p_value < alpha:
    print("Reject H₀{ha} → The average temperature is significantly greater than 30°C.")
else:
    print("accept null hypothesis{ho} → Not enough evidence that the average temperature is greater than 30°C.")



population_mean : 30
Sample Mean: 30.684999999999995
t-statistic: 5.600129086291858
p-value (one-tailed): 1.0594900976867397e-05
Reject H₀{ha} → The average temperature is significantly greater than 30°C.


## 4. **Two-Sample One-Tailed t-Test** (Right or Left) ➡️ ⬅️

**Used when:**  
You want to compare **two independent sample means** in a **specific direction** (greater or lesser).

### Right-Tailed Test (Greater Than):
**Example:**  
You want to test if the **average salary** of employees in Company A is **greater** than in Company B.

### Hypothesis:
- **Null Hypothesis (H₀):** Company A salary is **less than or equal to** Company B salary.  
  [
  H_0: mu_1 < mu_2
  ]
- **Alternative Hypothesis (H₁):** Company A salary is **greater** than Company B salary.  
  [
  H_1: mu_1 > mu_2
  ]

### Left-Tailed Test (Less Than):
**Example:**  
You want to test if the **average salary** of employees in Company A is **less** than in Company B.

### Hypothesis:
- **Null Hypothesis (H₀):** Company A salary is **greater than or equal to** Company B salary.  
  [
  H_0: mu_1 >mu_2
  ]
- **Alternative Hypothesis (H₁):** Company A salary is **less** than Company B salary.  
  [
  H_1: mu_1 < mu_2
  ]

---



In [4]:
company_a = [55, 60, 58, 62, 65, 59, 61, 64, 66, 63]
company_b = [52, 50, 49, 55, 53, 54, 51, 50, 52, 54]
import numpy as np
co_a_mean=np.mean(company_a)
co_b_mean=np.mean(company_b)

ho="Company A salary is less than or equal to Company B salary."
ha="Company A salary is greater than Company B salary"

# Perform two-sample one tailed t-test...
t_stat, p_two_tailed = stats.ttest_ind(company_a, company_b, equal_var=True)

# CONVERT P_VALUE_TWO_TAILED TO ONE TAIL P_VALUE:
p_value=p_two_tailed/2

# Print results
print("Company A Mean:", co_a_mean)
print("Company B Mean:", co_b_mean)
print("t-statistic:", t_stat)
print("p-value (one-tailed):", p_value)


# Significance level
alpha = 0.05

if p_value<alpha:
    print(f"reject null hypothesis:{ha}")
else:
    print(f"accept null hypothesis{ho}")    



Company A Mean: 61.3
Company B Mean: 52.0
t-statistic: 7.453927007776441
p-value (one-tailed): 3.3037709425884037e-07
reject null hypothesis:Company A salary is greater than Company B salary


## 5. **One-Sample Two-Tailed t-Test** 🔄

**Used when:**  
You want to test if the sample mean is **different** from the population mean in **both directions** (greater or less).

**Example:**  
You want to test if the average number of visitors to a website is **different** from 1,000.

### Hypothesis:
- **Null Hypothesis (H₀):** The mean number of visitors is 1,000.  
  [
  H_0: mu = 1000
  ]
- **Alternative Hypothesis (H₁):** The mean number of visitors is **not equal** to 1,000.  
  [
  H_1: mu != 1000
  ]

---



In [5]:
# ONE SAMPLE TWO TAILED TEST:

# Sample data: daily visitors to the website over 15 days
visitors = [980, 1020, 995, 1010, 990, 985, 1005, 1015, 970, 1000, 980, 1005, 995, 1010, 990]

# Population mean (assumed benchmark)
population_mean = 1000

# Hypotheses
ho = "Mean number of visitors = 1000"
ha = "Mean number of visitors != 1000"

# Perform one-sample t-test :
from scipy import stats
t_stat, p_value = stats.ttest_1samp(visitors,population_mean)
# P_VALUE IS TWO TAILED ALREADY...

# Significance level
alpha = 0.05

# Results
import numpy as np
print("Sample Mean:", np.mean(visitors))
print("t-statistic:", t_stat)
print("p-value:", p_value)

# Conclusion
if p_value < alpha:
    print("REJECT NULL HYPOTHESISH₀:", ha)
else:
    print("ACCEPT NULL HYPOTHESIS:", ho)


Sample Mean: 996.6666666666666
t-statistic: -0.8995824336398044
p-value: 0.383555035133315
ACCEPT NULL HYPOTHESIS: Mean number of visitors = 1000



---

### 🔍 **T-TEST KYA CHECK KARTA HAI?**

Hum dekh rahe hain ki:
- Sample ka average (mean) = **996.67**
- Population ka mean (jo humein diya gaya hai) = **1000**

Ab hum yeh test kar rahe hain:
> **Kya yeh 996.67 sirf ek chance se aaya hai ya yeh 1000 se *asli mein* alag hai?**

---

### 📌 **Null Hypothesis (H₀):**
"Sample mean aur population mean **same** hain"  
(yaani, 996.67 ≠ 1000 bas ek normal fluctuation hai)

### 📌 **Alternative Hypothesis (H₁):**
"Sample mean **different** hai population mean se"

---

### ⚖️ **T-test Kya Karta Hai?**
T-test check karta hai:
> "Kya yeh difference (996.67 vs 1000) itna bada hai ki use chance pe nahi chhod sakte?"

Agar **p-value < 0.05** hoti:
- Hum kehte: **yeh difference significant hai**  
  (matlab 996.67 sach mein 1000 se alag hai)

Lekin ab:
- **p-value = 0.359** → yeh **bahut zyada hai**  
- Iska matlab: yeh chhota difference (996.67 vs 1000) **normal variation ho sakta hai**  
- Isliye hum kehte hain:  
  ✅ **"Null hypothesis ko reject nahi kar sakte"**

---

### 📣 **Aakhri baat:**
> **996.67 ≠ 1000** hai, par yeh difference itna **bada nahi** hai ki hum kahein "yeh significant hai"

---


## 6. **Two-Sample Two-Tailed t-Test** 🔄🔄

**Used when:**  
You are comparing the **means** of **two independent samples** and testing if the means are **different** (either greater or less).

**Example:**  
You want to compare the **test scores** of students from two different schools and see if they differ.

### Hypothesis:
- **Null Hypothesis (H₀):** The average scores of the two schools are equal.  
  [
  H_0: mu_1 = mu_2
  ]
- **Alternative Hypothesis (H₁):** The average scores of the two schools are **not equal**.  
  [
  H_1: mu_1 != mu_2
  ]

---


In [6]:
# TWO SAMPLE TWO TAILED TEST:
# Test scores
school_a = [78, 85, 92, 88, 76, 81, 90, 84, 79, 87]
school_b = [72, 80, 78, 74, 77, 73, 76, 75, 79, 78]

import numpy as np
school_a_mean=np.mean(school_a)
school_b_mean=np.mean(school_b)

ho="school_a_mean = school_b_mean"
ha="school_a_mean != school_b_mean"

# perform TWO SAMPLE TWO TAILED TEST:
t_stat,p_value=stats.ttest_ind(school_a,school_b,equal_var=True)

# print details:
print(f"mean of school a test scores {school_a_mean}")
print(f"mean of school b test scores {school_b_mean}")
print(f"t_statistics : {t_stat}")
print(f"p_value(two tailed): {p_value}")

# Significance level
alpha = 0.05

# conclusion:
if p_value<alpha:
    print(f"reject null hypothesis {ha}")
else:
    print(f"accept null hypithesis {ho}")


mean of school a test scores 84.0
mean of school b test scores 76.2
t_statistics : 4.113500940517584
p_value(two tailed): 0.0006523312439442561
reject null hypothesis school_a_mean != school_b_mean


In [7]:
import pandas as pd



---

# 7. **Two-Proportion t-Test** 📊

### **Used When:**
You want to compare the **proportions** of two independent groups to see if they are **significantly different**.

**Example:**  
You want to compare the proportion of **male and female students** who passed an exam in two different classes.

### **Hypothesis:**
- **Null Hypothesis (H₀):** The proportion of male students who passed is equal to the proportion of female students who passed.  
  [
  H_0: p_1 = p_2
  ]
- **Alternative Hypothesis (H₁):** The proportion of male students who passed is not equal to the proportion of female students who passed.  
  [
  H_1: p_1 != p_2
  ]

---





### ✅ **Data**  
- Males: 60 students, 48 passed  
- Females: 70 students, 49 passed  


In [8]:
# TWO PROPORTION T-TEST:
# Data
success = [48, 49]        # number of students who passed
Tnos = [60, 70]           # total number of students


# Hypotheses
ho = "Proportion of males who passed = proportion of females who passed"
ha = "Proportion of males who passed ≠ proportion of females who passed"


# Perform two-proportion Z-test
from statsmodels.stats.proportion import proportions_ztest
z_stat, p_value = proportions_ztest(success, Tnos)

# Significance level
alpha = 0.05


# Results
print("Z-statistic:", z_stat)
print("p-value:", p_value)

# Conclusion
if p_value < alpha:
    print("Reject H₀:", ha)
else:
    print("Fail to reject H₀:", ho)

Z-statistic: 1.306030996358404
p-value: 0.19154201691201744
Fail to reject H₀: Proportion of males who passed = proportion of females who passed


# 10. **ANOVA Test (Analysis of Variance)** 📊

### **Used When:**
You want to compare the **means** of **three or more groups** to see if there is a significant difference in their means.

**Example:**  
You want to compare the **average test scores** of students in three different study programs (A, B, and C) to see if the means differ.

### **Hypothesis:**
- **Null Hypothesis (H₀):** The mean test scores for all study programs are equal.  
  [
  H_0: mu_1 = mu_2 = mu_3
  ]
- **Alternative Hypothesis (H₁):** At least one of the mean test scores is different.  
  [
  H_1: At least one mu  is different.
  ]

**Note:**  
ANOVA doesn't tell you which groups are different—just that there is a difference. If **H₁** is true, post-hoc tests (e.g., Tukey's HSD) are often conducted to determine which groups differ.

---



In [9]:
# ANOVA ANALYSIS OF VARIENCE:
# Scores from 3 different study programs
program_A = [85, 88, 90, 92, 87]
program_B = [78, 74, 80, 79, 77]
program_C = [90, 92, 88, 91, 89]

# Hypotheses
ho = "The mean test scores for all study programs are equal."
ha = "At least one of the mean test scores is different."

# Perform one-way ANOVA
from scipy import stats
f_stat, p_value = stats.f_oneway(program_A, program_B, program_C)

# Significance level
alpha = 0.05

# Results
print("F-statistic:", f_stat)
print("p-value:", p_value)

# Conclusion
if p_value < alpha:
    print("Reject H₀:", ha)
else:
    print("Fail to reject H₀:", ho)

F-statistic: 45.192052980132374
p-value: 2.5923415606052823e-06
Reject H₀: At least one of the mean test scores is different.


In [10]:
# Combine all scores and group labels:
scores = [85, 88, 90, 92, 87, 78, 74, 80, 79, 77, 90, 92, 88, 91, 89]
groups = ['A']*5 + ['B']*5 + ['C']*5

# Create DataFrame
import pandas as pd
df = pd.DataFrame({'score': scores, 'group': groups})

# Tukey HSD test
from statsmodels.stats.multicomp import pairwise_tukeyhsd
tukey_result = pairwise_tukeyhsd(endog=df['score'], groups=df['group'], alpha=0.05)
print(tukey_result)


 Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower    upper  reject
-----------------------------------------------------
     A      B    -10.8    0.0 -14.5855 -7.0145   True
     A      C      1.6 0.5162  -2.1855  5.3855  False
     B      C     12.4    0.0   8.6145 16.1855   True
-----------------------------------------------------


In [11]:
df

score group
0      85     A
1      88     A
2      90     A
3      92     A
4      87     A
5      78     B
6      74     B
7      80     B
8      79     B
9      77     B
10     90     C
11     92     C
12     88     C
13     91     C
14     89     C

# 11. **Chi-Square Test** 🔢

### **Used When:**
You want to test the **association** between two **categorical variables** or compare the **observed** and **expected frequencies** in a contingency table.

There are two main types of Chi-Square tests:
1. **Chi-Square Goodness of Fit Test:** Tests if an observed distribution matches a specific distribution.
2. **Chi-Square Test of Independence:** Tests if two categorical variables are independent.

### **Example 1:** **Chi-Square Goodness of Fit**  
You want to test if the distribution of **coin flips** (heads or tails) is **fair** (50% heads, 50% tails).

### **Hypothesis:**
- **Null Hypothesis (H₀):** The coin is fair (heads and tails are equally likely).  
  [
  H_0: heads = tails = 0.5
  ]
- **Alternative Hypothesis (H₁):** The coin is **not fair** (the probabilities are not equal).  
  [
  H_1: heads != tails != 0.5
  ]



In [18]:
'''
🔹 1. Chi-Square Goodness of Fit Test
You compare observed frequencies with expected frequencies.

🪙 Example: Fair Coin Toss
Suppose you flip a coin 100 times: 55 heads, 45 tails.
'''
ho = "head and tail frequencies are equal"
ha = "head and tail frequency are not equal"

# Observed frequencies
observed = [55, 45]

# Expected frequencies if fair coin
expected = [50, 50]

# Perform Chi-Square Goodness of Fit Test
from scipy.stats import chisquare
chi2_stat, p_val = chisquare(f_obs=observed, f_exp=expected)

print("Chi-square Statistic:", chi2_stat)
print("p-value:", p_val)

alpha = 0.05

# compare p_val with alpha:
if p_val < alpha:
    print(f"Reject null hypothesis: {ha}")
else:
    print(f"Accept null hypothesis: {ho}")
    
'''
✅ Output Meaning:
If p_val < 0.05: Significant result → coin might not be fair.

If p_val ≥ 0.05: No significant evidence → coin is likely fair.'''        


Chi-square Statistic: 1.0
p-value: 0.31731050786291115
Accept null hypothesis: head and tail frequencies are equal


'\n✅ Output Meaning:\nIf p_val < 0.05: Significant result → coin might not be fair.\n\nIf p_val ≥ 0.05: No significant evidence → coin is likely fair.'

### **Example 2:** **Chi-Square Test of Independence**  
You want to test if there is an association between **gender** (male/female) and **job satisfaction** (satisfied/unsatisfied) in a company.

### **Hypothesis:**
- **Null Hypothesis (H₀):** There is no association between gender and job satisfaction (they are independent).  
  [
  H_0:Gender and Job Satisfaction are independent.
  ]
- **Alternative Hypothesis (H₁):** There is an association between gender and job satisfaction (they are not independent).  
  [
  H_1:Gender and Job Satisfaction are dependent.
  ]

---

In [24]:
# Hypotheses
ho = "Gender and Job Satisfaction are independent"
ha = "Gender and Job Satisfaction are dependent"

# Contingency Table:
#                Satisfied  Unsatisfied
#    Male            40          10
#    Female          30          20


data = np.array([[40, 10],
                 [30, 20]])

# Perform Chi-Square Test of Independence
from scipy.stats import chi2_contingency
chi2_stat, p_val, dof, expected = chi2_contingency(data)# double dimension numpy array


print("Chi-square Statistic:", chi2_stat)
print("p-value:", p_val)
print("Degrees of Freedom:", dof)
print("Expected Frequencies:\n", expected)


# Significance Level
alpha = 0.05

# Decision
if p_val < alpha:
    print(f"Reject null hypothesis: {ha}")
else:
    print(f"Accept null hypothesis: {ho}")




Chi-square Statistic: 3.8571428571428577
p-value: 0.04953461343562649
Degrees of Freedom: 1
Expected Frequencies:
 [[35. 15.]
 [35. 15.]]
Reject null hypothesis: Gender and Job Satisfaction are dependent
